src/utils/actions.js

```diff
@@ -0,0 +1,254 @@
+            });
+            filteredResources.push.apply(filteredResources, dateFiltered);
+          } else {
+            // path or searchParam are required
+            const dateFiltered = resources.filter(resource => {
+              // get value by param
+              const date = getBySearchParam(resource, searchParam); // assume its a datetime?
+              return compareDates(date, filterPeriod);
+            });
+            filteredResources.push.apply(filteredResources, dateFiltered);
+          }
+        } else {
+          // if there's no date filter, we can just push all the resources
+          // in to be filtered by code
+          filteredResources.push.apply(filteredResources, resources);
+        }
+
+        const codeFilter = input.codeFilter;
+        if (codeFilter) {
+          // code filter isn't under the same constraint as datefilter
+          // we can assume that it's just a path
+          const path = codefilter.path;
+          const searchParam = codeFilter.searchParam;
+          if (path) {
+            // no need to refilter resources
+            const filteredResources = filteredResources.filter(resource => {
+              const code = getByPath(resource, path);
+              return compareCodes(code, codeFilter.code); // codeFilter.code is of type Coding, could also be a valueSet?
+            });
+          } else if (searchParam) {
+            const filteredResources = filteredResources.filter(resource => {
+              const code = getBySearchParam(resource, searchParam);
+              return compareCodes(code, codeFilter.code);
+            });
+          }
+        }
+        return context.records.push(filteredResources);
+      });
+    }
+  },
+  'execute-research-query': context => {
+    // noop
+  }
+};
+
+function validateProfile(report, structureDefinition) {
+  // TODO: Perhaps this is something one of the servers/clients would handle?
+  return report;
+}
+
+function getByPath(resource, path) {
+  // TODO: function to search resources by path
+  // and return correct attribute
+  return resource;
+}
+
+function compareCodes(coding1, coding2) {
+  // assume they're both coding
+  return coding1.code === coding2.code && coding1.system === coding2.system;
+}
+
+function getBySearchParam(resource, searchParam) {
+  // TODO: function to search resources by param
+  // and return correct attribute
+  return resource.date;
+}
+
+function compareDates(period1, period2) {
+  // TODO: returns true if the periods overlap
+}
+function createBundle(records) {
+  const bundle = {
+    resourceType: 'Bundle',
+    type: 'collection',
+    entry: []
+  };
+
+  records.forEach(record => {
+    bundle.push({
+      resource: record
+    });
+  });
+
+  return bundle;
+}
+
+function getPatient(context) {
+  const patientId = context.patient;
+  const patient = context.client.source.read(patientId);
+  context.patient = patient;
+  return context;
```

Comment:
Is there a reason `context.patient` is just the identifier but `context.encounter` is the Encounter? Unless there's a good reason I propose changing to make the `context.patient` the entire Patient resource. It is then the responsibility of the subscription notification handler to retrieve the patient while initializing the context.

If it should remain as just the identifier, wouldn't the patient exist in `context.records`? So it could be found from there instead of making a network call.

Reply:
Yeah we should be consistent if possible. Either both should be identifiers or both should be resources, whichever makes more sense

Reply:
Looks like `context.patient` isn't even used anymore in the latest commit. Based on the tests though it has been updated to be the complete patient instead of just the id

Reply:
The patient will probably be needed for `check-participant-registration` but I'm still not quite sure what that actions implementation is meant to be